File: backend/internal/domain/savings.go
package domain

import "time"

type SavingsType struct {
	ID            string    `json:"id"`
	Name          string    `json:"name"`
	DefaultAmount string    `json:"defaultAmount"`
	IsMandatory   bool      `json:"isMandatory"`
	CreatedAt     time.Time `json:"createdAt"`
}

type SavingsTransaction struct {
	ID            string    `json:"id"`
	MemberID      string    `json:"memberId"`
	SavingsTypeID string    `json:"savingsTypeId"`
	Type          string    `json:"type"`
	Amount        string    `json:"amount"`
	BalanceAfter  string    `json:"balanceAfter"`
	Note          *string   `json:"note"`
	ReceiptNumber string    `json:"receiptNumber"`
	CreatedBy     string    `json:"createdBy"`
	CreatedAt     time.Time `json:"createdAt"`
}

type SavingsRepository interface {
	GetTypes() ([]SavingsType, error)
	GetTransactions(memberID string) ([]SavingsTransaction, error)
	CreateTransaction(input CreateSavingsInput) (*SavingsTransaction, error)
}

type SavingsService interface {
	GetTypes() ([]SavingsType, error)
	GetTransactions(memberID string) ([]SavingsTransaction, error)
	Deposit(input CreateSavingsInput) (*SavingsTransaction, error)
	Withdraw(input CreateSavingsInput) (*SavingsTransaction, error)
}

type CreateSavingsInput struct {
	MemberID      string  `json:"memberId"`
	SavingsTypeID string  `json:"savingsTypeId"`
	Type          string  `json:"type"`
	Amount        string  `json:"amount"`
	BalanceAfter  string  `json:"balanceAfter"`
	Note          *string `json:"note"`
	ReceiptNumber string  `json:"receiptNumber"`
	CreatedBy     string  `json:"createdBy"`
}

File: data-service/src/db/schema/savings.ts
import { pgTable, uuid, varchar, numeric, boolean, timestamp, text } from 'drizzle-orm/pg-core'
import { members } from './members.js'

export const savingsTypes = pgTable('savings_types', {
  id: uuid('id').primaryKey().defaultRandom(),
  name: varchar('name', { length: 50 }).notNull(),
  defaultAmount: numeric('default_amount', { precision: 15, scale: 2 }).notNull(),
  isMandatory: boolean('is_mandatory').notNull(),
  createdAt: timestamp('created_at').defaultNow().notNull(),
})

export const savingsTransactions = pgTable('savings_transactions', {
  id: uuid('id').primaryKey().defaultRandom(),
  memberId: uuid('member_id').references(() => members.id).notNull(),
  savingsTypeId: uuid('savings_type_id').references(() => savingsTypes.id).notNull(),
  type: varchar('type', { length: 10 }).notNull(),
  amount: numeric('amount', { precision: 15, scale: 2 }).notNull(),
  balanceAfter: numeric('balance_after', { precision: 15, scale: 2 }).notNull(),
  note: varchar('note', { length: 255 }),
  receiptNumber: varchar('receipt_number', { length: 50 }).unique().notNull(),
  createdBy: text('created_by').notNull(),
  createdAt: timestamp('created_at').defaultNow().notNull(),
})
